crates/telemetry/src/detector.rs

```diff
@@ -0,0 +1,47 @@
+use std::{env, time::Duration};
+
+use opentelemetry::{Key, KeyValue, Value};
+use opentelemetry_sdk::{
+    resource::{EnvResourceDetector, ResourceDetector},
+    Resource,
+};
+
+const OTEL_SERVICE_NAME: &str = "OTEL_SERVICE_NAME";
+
+/// Custom resource detector for Spin relevant attributes service.name and service.version.
+///
+/// To set service.name this detector will first try `OTEL_SERVICE_NAME` env. If it's not available,
+/// then it will check the `OTEL_RESOURCE_ATTRIBUTES` env and see if it contains `service.name`
+/// resource. If it's also not available, it will use `spin`.
+///
+/// To set service.version, it will use the spin_version passed in new.
+#[derive(Debug)]
+pub struct SpinResourceDetector {
+    spin_version: String,
+}
+
+impl SpinResourceDetector {
+    /// Create a new instance of SpinResourceDetector.
+    pub fn new(spin_version: String) -> Self {
+        SpinResourceDetector { spin_version }
+    }
+}
+
+impl ResourceDetector for SpinResourceDetector {
```

Comment:
This seems to overlap with `EnvResourceDetector`. Is the point here just to set the default `service.name = spin`? If so, I suspect you could get the same effect with less code by manually creating a Resource in `otel_tracing_layer` and then [`merge`](https://docs.rs/opentelemetry_sdk/0.22.1/opentelemetry_sdk/struct.Resource.html#method.merge)ing in the detected resource attrs.

Reply:
It also sets the version.

That's a good idea, but I think I prefer keeping it in the detector. Inlining it would remove the need for a struct definition and a constructor, but otherwise most of the code would remain. I prefer following the pattern of a detector.

Reply:
That's fine - and this certainly isn't a blocking comment - but if its effectively setting a default and a hard-coded version I might be tempted to argue that it isn't "detecting" anything. :slightly_smiling_face: